rep-0151.rst

```diff
@@ -0,0 +1,260 @@
+Since then it was decided the first ROS 1 distribution to support Python 3
+will be Noetic Ninjemys (May 2020) [3]_, and it will only support Python 3.
+This REP only describes the current proposal since the change in requirements
+has made past proposals obsolete.
+
+This proposal does not specify a minimum minor version for Python.
+The minimum version will being defined in REP 3 [4]_ together with all other
+versions.
+
+Proposal
+========
+
+Python 3 support at the time of Noetic release means:
+
+* All ROS tooling supports creating a Python 3 only ROS distro
+* All ROS packages released to Noetic support Python 3
+
+The amount of work to transition to Python 3 is expected to be significant.
+The entire ROS community will need to share the workload.
+This effort should be started as soon as possible to reduce the risk of the
+Noetic release being delayed.
+
+Tooling support for Python 3
+----------------------------
+
+rosdep keys
+'''''''''''
+All exsiting Python rosdep keys including ``<platform>_python3`` entries will
+stay the same to avoid breaking unknown existing use cases.
+Even if an key refers to a Python 2 package on one platform and Python 3 on
+another, it should stay that way.
+
+Some Python rosdep keys have ``<platform>_python3`` entries.
+This was an experiment at having a rosdep key conditionally evaluate to either
+a Python 2 or Python 3 system dependency.
+No new ``<platform>_python3`` entries should be added, and the entries should
+not be relied on.
+
+For new Python keys, when a platform has separate Python 2 and Python 3 versions
+of a package there should be a rosdep key for each.
+For example, there are two keys ``python-catkin-pkg`` and ``python3-catkin-pkg``.
+The ``python-catkin-pkg`` key should resolve to a Python 2 dependency on any
+platform where one exists, while ``python3-catkin-pkg`` should only resolve to a
+Python 3 dependency.
+
+
+ROS_PYTHON_VERSION
+''''''''''''''''''
+
+``ROS_PYTHON_VERSION`` is an environment variable that is set to either ``2`` or
+``3`` to indicate the major version of Python supported by a ROS distro.
+The intent is to allow ``package.xml`` files to have conditional Python
+dependencies.
+
+It is set by the package ``ros_environment`` when sourcing a ROS workspace.
+The version set will be dictated by REP 3 [4]_.
+For example, in Melodic it defaults to ``2``, but in Noetic this must be ``3``.
+
+Users building ROS from source will not have an existing ROS workspace set, so
+the variable won't be set and conditional dependencies can't be evalulated
+when using ``rosdep`` to instal them.
+To fix this, ``rosdep`` should set ``ROS_PYTHON_VERSION`` to the same version of
+Python used to invoke it.
+It will get this information using ``sys.version[0]``.
+A user wanting to use a different Python version must set ``ROS_PYTHON_VERSION``
+explicitly.
+
+rosdep and pip packages
+'''''''''''''''''''''''
+Unlike debian packages, keys that resolve to packages on PyPI may refer to
```

Comment:
```suggestion
Unlike Debian packages, keys that resolve to packages on PyPI may refer to
```
As far as I can seen, just by looking around online, Debian is always capitalized. Ignore if you know otherwise.